design-proposals/klip-8-interactive-queries.md

```diff
@@ -0,0 +1,143 @@
+
+* All aggregates supported by continuous queries should be supported by interactive queries.
+* Examples:
+  * `SELECT count(*) FROM table WHERE ROWKEY < 10`
+  * `SELECT day(hour), sum(count) FROM table GROUP BY day`
+  * `SELECT sum(count) FROM table WHERE ts >= ‘2019-01-01’ AND ts < ‘2019-01-02’`
+
+### Tier 4 - Joins
+
+Joins are also fundamental to deeply analyzing datasets and the relationships between various entities. They will also likely be the most complex to support in a clustered environment, which is why they’ve been put in a separate and final tier. As with aggregates, users may already use joins in their materialized view definitions, which may partially mitigate the cost of their absence within the context of interactive queries.
+
+For these reasons, we should probably not focus much on joins until the previous tiers have been delivered.
+
+* Without relatively robust indexing, it will probably only be initially feasible to join on row keys.
+* Examples:
+  * `SELECT … FROM t0 JOIN t1 ON t0.ROWKEY = t1.ROWKEY …`
+  * `SELECT stats_1m.count AS last_minute, stats_60m.count AS last_hour FROM stats_1m JOIN stats_60m ON stats_1m.ROWKEY = stats_1h.ROWKEY`
+
+## User experience
+
+The user experience for interactive queries should be quite simple: if a user issues a query against a KSQL table, that query should return a finite set of rows as quickly as possible.
+
+Additionally:
+
+* All interactive queries must be issued via the REST API.
+* Any interactive query can target any KSQL server.
+* Key movement should be handled transparently: if a query launches and a targeted key physically moves, KSQL should redirect the query to the appropriate KSQL server to obtain the correct result.
+* For interactive queries against windowed queries, the user may specify a time range via the `WINDOWSTART` and `WINDOWEND` psuedocolumns. For example, `SELECT * FROM windowed_aggregate WHERE ROWKEY = 'rowkey' AND WINDOWSTART >= start_ts AND WINDOWEND < end_ts`
+
+## Open questions
+
+1. How will KSQL differentiate between interactive and streaming queries? Should the same differentiation mechanism be applied to both transient queries and CTAS/CSAS statements?
+
+2. Do we want to support interactive queries on streams? An interactive query on a stream would execute on a stream, beginning at some offset, run until completion and return the final result. In contrast to a streaming query, an interactive query on a stream would not produce incremental updates as the query progresses over the input. Here is a matrix of all possible query types after the addition of interactive queries:
```

Comment:
My 2c. That would introduce a lot of seeking and resetting on the kafka cluster and may not be a good thing to support. Are there are concrete use-cases for this, we have heard about?

Reply:
I agree that interactive queries on streams probably shouldn't be a high priority. I see their utility in theory, but I'm not aware of any actual user demand for them at this point.

Reply:
I'd certainly keep these out of scope for this piece of work. I see these as more _temporal_ queries, which build aggregates across historic data sets.  There is _totally_ a use-case for such patterns, but they are a different beast to the _interactive_ queries covered by this KLIP.